### PR TITLE
[FLINK-9034] [FLINK-9035] [core] Fix state descriptors

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

Fixes two issue with the `StateDescriptors` that are used to obtain state access in transformation functions: 

### Broken Equals and hashCode

`equals()` and `hashCode()` depends on fields that are not always set and that may change during the life of a state descriptor. That is especially problematic, because the state descriptors are keys in a map, and if the meaning of `equals()` and `hashCode()` changes after insertion, the objects become keys that cannot be references / matched.

This pull request changes `equals()` and `hashCode()` to only take state name and descriptor type (by class) into account for hashCode and equality, which are always constant and not changing as part of serializer initialization.

**Illustration of the problem:**

The following code fails with a `NullPointerException`, because the `hashCode()` method tries to access the serializer field, which may be uninitialized at that point.

```java
ValueStateDescriptor<String> descr = new ValueStateDescriptor<>("name", String.class);
descr.hashCode(); // exception
```

The equals() method is equally broken (no pun intended):
```java
ValueStateDescriptor<String> a = new ValueStateDescriptor<>("name", String.class);
ValueStateDescriptor<String> b = new ValueStateDescriptor<>("name", String.class);

a.equals(b) // exception
b.equals(a) // exception
a.initializeSerializerUnlessSet(new ExecutionConfig());
a.equals(b) // false
b.equals(a) // exception
b.initializeSerializerUnlessSet(new ExecutionConfig());
a.equals(b) // true
b.equals(a) // true
```

### Type Information dropped prematurely

The following code is currently problematic:
```java
public class MyFunction extends RichMapFunction<A, B>  {

    private final ValueStateDescriptor<MyType> descr = new ValueStateDescriptor<>("state name", MyType.class);

    private ValueState<MyType> state;

    @Override
    public void open(Configuration cfg) {
        state = getRuntimeContext().getValueState(descr);
    }
}
```

The problem is that the state descriptor drops the type information and creates a serializer before serialization as part of shipping the function in the cluster. To do that, it initializes the serializer with an empty execution config, making serialization inconsistent.

This is mainly an artifact from the days when dropping the type information before shipping was necessary, because the type info was not serializable. It now is, and we can fix that bug.

## Verifying this change

**This change is sensitive, because it touches the structures that all users use to obtain access to persistent state.**

This change adds a series of Unit tests to validate the fixed behavior.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no)**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no) - **All changes should preserve full API compatibility.**
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know): **Touches the structures that give access to checkpointed state.**
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java [hunk 1]
```diff
@@ -77,18 +80,22 @@
 
 	/** The serializer for the type. May be eagerly initialized in the constructor,
 	 * or lazily once the type is serialized or an ExecutionConfig is provided. */
+	@Nullable
 	protected TypeSerializer<T> serializer;
 
+	/** The type information describing the value type. Only used to lazily create the serializer
+	 * and dropped during serialization */
+	@Nullable
```
Comment (sihuazhou) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java]: nit:Type information will not dropped during serialization now, it dropped in `initializeSerializerUnlessSet()`.

File: flink-core/src/test/java/org/apache/flink/api/common/state/AggregatingStateDescriptorTest.java [hunk 2]
```diff
@@ -41,16 +40,11 @@
...
 	@Test
 	public void testSerializerDuplication() {
-		TypeSerializer<Long> serializer = mock(TypeSerializer.class);
-		when(serializer.duplicate()).thenAnswer(new Answer<TypeSerializer<Long>>() {
-			@Override
-			public TypeSerializer<Long> answer(InvocationOnMock invocation) throws Throwable {
-				return mock(TypeSerializer.class);
-			}
-		});
+		// we need a serializer that actually duplicates for testing (a stateful one)
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/common/state/AggregatingStateDescriptorTest.java]: Will this condition always hold? Should we maybe guard this assumption with an assertion, i.e. assert that the result of `serialiser.duplicate()` is different from the original serialiser?

File: flink-core/src/test/java/org/apache/flink/api/common/state/ListStateDescriptorTest.java [hunk 3]
```diff
@@ -121,13 +117,9 @@ public void testValueStateDescriptorAutoSerializer() throws Exception {
...
 	@Test
 	public void testSerializerDuplication() {
-		TypeSerializer<String> statefulSerializer = mock(TypeSerializer.class);
-		when(statefulSerializer.duplicate()).thenAnswer(new Answer<TypeSerializer<String>>() {
-			@Override
-			public TypeSerializer<String> answer(InvocationOnMock invocation) throws Throwable {
-				return mock(TypeSerializer.class);
-			}
-		});
+		// we need a serializer that actually duplicates for testing (a stateful one)
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/common/state/ListStateDescriptorTest.java]: Same comment as above.

File: flink-core/src/test/java/org/apache/flink/api/common/state/MapStateDescriptorTest.java [hunk 4]
```diff
@@ -129,23 +125,12 @@ public void testMapStateDescriptorAutoSerializer() throws Exception {
...
-				return mock(TypeSerializer.class);
-			}
-		});
-		when(valueSerializer.duplicate()).thenAnswer(new Answer<TypeSerializer<Long>>() {
-			@Override
-			public TypeSerializer<Long> answer(InvocationOnMock invocation) throws Throwable {
-				return mock(TypeSerializer.class);
-			}
-		});
+		// we need a serializer that actually duplicates for testing (a stateful one)
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/common/state/MapStateDescriptorTest.java]: See above

File: flink-core/src/test/java/org/apache/flink/api/common/state/ReducingStateDescriptorTest.java [hunk 5]
```diff
@@ -118,17 +115,14 @@ public void testValueStateDescriptorAutoSerializer() throws Exception {
...
-			@Override
-			public TypeSerializer<String> answer(InvocationOnMock invocation) throws Throwable {
-				return mock(TypeSerializer.class);
-			}
-		});
-
-		ReduceFunction<String> reducer = mock(ReduceFunction.class);
-
-		ReducingStateDescriptor<String> descr = new ReducingStateDescriptor<>("foobar", reducer, statefulSerializer);
+		// we need a serializer that actually duplicates for testing (a stateful one)
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/common/state/ReducingStateDescriptorTest.java]: See above

File: flink-core/src/test/java/org/apache/flink/api/common/state/ValueStateDescriptorTest.java [hunk 6]
```diff
@@ -149,13 +145,9 @@ public void testVeryLargeDefaultValue() throws Exception {
...
 	@Test
 	public void testSerializerDuplication() {
-		TypeSerializer<String> statefulSerializer = mock(TypeSerializer.class);
-		when(statefulSerializer.duplicate()).thenAnswer(new Answer<TypeSerializer<String>>() {
-			@Override
-			public TypeSerializer<String> answer(InvocationOnMock invocation) throws Throwable {
-				return mock(TypeSerializer.class);
-			}
-		});
+		// we need a serializer that actually duplicates for testing (a stateful one)
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/common/state/ValueStateDescriptorTest.java]: See above

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java [hunk 7]
```diff
@@ -77,18 +80,22 @@
 
 	/** The serializer for the type. May be eagerly initialized in the constructor,
 	 * or lazily once the type is serialized or an ExecutionConfig is provided. */
+	@Nullable
 	protected TypeSerializer<T> serializer;
 
+	/** The type information describing the value type. Only used to lazily create the serializer
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java]: nit: I think this was also a copying error in the original comment, but this is not necessarily a "value", unless we simply thing of all state as a value, in which case I'm fine with this.

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java [hunk 8]
```diff
@@ -249,12 +257,13 @@ public boolean isSerializerInitialized() {
 	 */
 	public void initializeSerializerUnlessSet(ExecutionConfig executionConfig) {
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java]: This is slightly orthogonal to this change, but: could we get rid of this method and instead change `getSerializer()` to `getSerializer(ExecutionConfig)`. That way, we don't have to be concerned about forgetting to call `initializeSerializerUnlessSet()`.

File: flink-core/src/test/java/org/apache/flink/api/common/state/StateDescriptorTest.java [hunk 9]
```diff
@@ -130,6 +135,31 @@ public void testInitializeSerializerAfterSerializationWithCustomConfig() throws
...
+	//  Tests for serializer initialization
+	// ------------------------------------------------------------------------
+
+	/**
+	 * FLINK-6775, tests that the returned serializer is duplicated.
+	 * This allows to share the state descriptor across threads.
+	 */
+	@Test
+	public void testSerializerDuplication() throws Exception {
+		// we need a serializer that actually duplicates for testing (a stateful one)
```
Comment (aljoscha) [inline flink-core/src/test/java/org/apache/flink/api/common/state/StateDescriptorTest.java]: Same as above, we should assert that assumption

**Comment by StephanEwen** [comment]
Concerning guarding the test assumptions: I think it is fine, because if the KryoSerializer is not duplicating properly any more (assumption violated), the test will also fail. (The problematic case would be the other way around, a serializer that does not duplicate and we check that the same instance is returned).

@aljoscha Do you think this should go into 1.5 as well?

@aljoscha Concerning changing `initializeSerializerUnlessSet(ExecutionConfig)` to `getSerializer(ExecutionConfig)` - I think that is a good idea, but would do that in a separate step. We need to re-work anyways a bit how we store state descriptors and obtain and reconfigure serializers.

**Comment by aljoscha** [comment]
@StephanEwen Yes, I think this should go into 1.5.0 because it fixes potential (and real) problems. And yes, I wasn't suggesting to remove `initializeSerializerUnlessSet(ExecutionConfig)` now, but it seemed like a good place to mention it. 😃

**Comment by StephanEwen** [comment]
Manually merged in f3a519712fb31f7b71181e876c3c3d5fff08eb71

**Pull request closed without merging**